File: building_block_apps/kafka/java/Producer.java
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.apache.kafka.clients.producer.Callback;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Kafka Producer Example in Java
 * 
 * This class demonstrates how to produce messages to a Kafka topic using the Kafka Java client.
 * It includes error handling, delivery confirmation, and performance monitoring.
 * 
 * Usage:
 *     java Producer [--topic TOPIC] [--count COUNT] [--rate RATE]
 * 
 * Example:
 *     java Producer --topic my-topic --count 1000 --rate 10
 */
public class Producer {
    private static final Logger logger = LoggerFactory.getLogger(Producer.class);
    private final String topic;
    private final KafkaProducer<String, String> producer;
    private final KafkaCommon.MessageStats stats;
    private final AtomicBoolean running;
    
    public Producer(String topic) {
        this.topic = topic;
        this.stats = new KafkaCommon.MessageStats();
        this.running = new AtomicBoolean(true);
        
        // Setup shutdown hook for graceful shutdown
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            logger.info("🛑 Shutdown signal received, stopping gracefully...");
            running.set(false);
        }));
        
        // Create producer with configuration
        Properties config = KafkaCommon.getProducerConfig();
        this.producer = new KafkaProducer<>(config);
        
        logger.info("🔌 Connected to Kafka brokers: {}", KafkaCommon.getBootstrapServers());
        logger.info("✅ Producer initialized successfully!");
    }
    
    /**
     * Send a message to the Kafka topic with callback handling.
     * 
     * @param key Message key (for partitioning)
     * @param message Message content
     * @param latch CountDownLatch for synchronization
     */
    public void sendMessage(String key, String message, CountDownLatch latch) {
        try {
            ProducerRecord<String, String> record = new ProducerRecord<>(topic, key, message);
            
            producer.send(record, new Callback() {
                @Override
                public void onCompletion(RecordMetadata metadata, Exception exception) {
                    try {
                        if (exception == null) {
                            // Success
                            stats.recordSent();
                            logger.info("✅ Message sent to {}[{}] at offset {}", 
                                      metadata.topic(), metadata.partition(), metadata.offset());
                        } else {
                            // Error
                            stats.recordError();
                            logger.error("❌ Message delivery failed: {}", exception.getMessage());
                        }
                    } finally {
                        latch.countDown();
                    }
                }
            });
            
        } catch (Exception e) {
            logger.error("❌ Error sending message: {}", e.getMessage());
            stats.recordError();
            latch.countDown();
        }
    }
    
    /**
     * Produce a specified number of messages.
     * 
     * @param count Number of messages to produce
     * @param rate Messages per second (0 = no limit)
     */
    public void produceMessages(int count, double rate) {
        logger.info("🚀 Starting to produce {} messages to topic '{}'", count, topic);
        if (rate > 0) {
            logger.info("📊 Rate limit: {} messages/second", rate);
        } else {
            logger.info("📊 Rate limit: unlimited");
        }
        
        long sleepTimeMs = rate > 0 ? (long) (1000.0 / rate) : 0;
        CountDownLatch latch = new CountDownLatch(count);
        
        for (int i = 0; i < count && running.get(); i++) {
            // Create sample message
            String messageData = KafkaCommon.createSampleMessage(i, "producer_demo");
            String messageKey = "key_" + i;
            
            // Send message
            sendMessage(messageKey, messageData, latch);
            logger.info("📤 Sent message {}/{}: {}", i + 1, count, messageKey);
            
            // Rate limiting
            if (sleepTimeMs > 0) {
                try {
                    Thread.sleep(sleepTimeMs);
                } catch (InterruptedException e) {
                    logger.warn("Sleep interrupted, stopping...");
                    Thread.currentThread().interrupt();
                    break;
                }
            }
            
            // Print stats every 100 messages
            if ((i + 1) % 100 == 0) {
                stats.printStats();
            }
        }
        
        // Wait for all messages to be delivered
        logger.info("🔄 Waiting for all messages to be delivered...");
        try {
            latch.await();
        } catch (InterruptedException e) {
            logger.warn("Wait interrupted");
            Thread.currentThread().interrupt();
        }
        
        // Flush remaining messages
        logger.info("🔄 Flushing remaining messages...");
        producer.flush();
        
        logger.info("✅ Finished producing messages!");
        stats.printStats();
    }
    
    /**
     * Close the producer connection.
     */
    public void close() {
        if (producer != null) {
            logger.info("🔌 Closing producer connection...");
            producer.close();
            logger.info("✅ Producer closed successfully");
        }
    }
    
    /**
     * Main method to run the Kafka producer.
     * 
     * @param args Command line arguments
     */
    public static void main(String[] args) {
        KafkaCommon.Args parsedArgs = new KafkaCommon.Args(args);
        
        // Check for help flag
        if (parsedArgs.hasFlag("--help")) {
            parsedArgs.printUsage("Producer");
            return;
        }
        
        // Parse arguments
        String topic = parsedArgs.get("--topic", KafkaCommon.getDefaultTopic());
        int count = parsedArgs.getInt("--count", 10);
        double rate = parsedArgs.getDouble("--rate", 1.0);
        
        System.out.println("🎯 Kafka Producer Example");
        System.out.println("=".repeat(50));
        System.out.println("📌 Topic: " + topic);
        System.out.println("📊 Message count: " + count);
        System.out.println("⚡ Rate: " + rate + " messages/second");
        System.out.println("=".repeat(50));
        
        Producer producer = new Producer(topic);
        
        try {
            // Produce messages
            producer.produceMessages(count, rate);
            
        } catch (Exception e) {
            logger.error("❌ Unexpected error: {}", e.getMessage(), e);
            System.exit(1);
        } finally {
            producer.close();
        }
    }
}

File: building_block_apps/kafka/python/producer.py
#!/usr/bin/env python3
"""
Kafka Producer Example in Python

This script demonstrates how to produce messages to a Kafka topic using the kafka-python library.
It includes error handling, delivery confirmation, and performance monitoring.

Usage:
    python producer.py [--topic TOPIC] [--count COUNT] [--rate RATE]

Example:
    python producer.py --topic my-topic --count 1000 --rate 10

Requirements:
    pip install kafka-python
"""

import argparse
import signal
import sys
import time
from typing import Optional

try:
    from kafka import KafkaProducer
    from kafka.errors import KafkaError
except ImportError:
    print("❌ Error: kafka-python library not found!")
    print("📦 Install with: pip install kafka-python")
    sys.exit(1)

from kafka_common import (
    get_producer_config, create_sample_message, serialize_message,
    get_default_topic, MessageStats
)


class EnhancedKafkaProducer:
    """Enhanced Kafka producer with monitoring and error handling."""
    
    def __init__(self, topic: str):
        self.topic = topic
        self.producer: Optional[KafkaProducer] = None
        self.stats = MessageStats()
        self.running = True
        
        # Setup signal handlers for graceful shutdown
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)
    
    def _signal_handler(self, signum, frame):
        """Handle shutdown signals gracefully."""
        print(f"\n🛑 Received signal {signum}, shutting down gracefully...")
        self.running = False
    
    def connect(self) -> bool:
        """
        Connect to Kafka cluster.
        
        Returns:
            True if connection successful, False otherwise
        """
        try:
            config = get_producer_config()
            print(f"🔌 Connecting to Kafka brokers: {config['bootstrap_servers']}")
            
            # Create producer with string serialization
            self.producer = KafkaProducer(
                bootstrap_servers=config['bootstrap_servers'],
                acks=config['acks'],
                retries=config['retries'],
                retry_backoff_ms=config['retry_backoff_ms'],
                request_timeout_ms=config['request_timeout_ms'],
                delivery_timeout_ms=config['delivery_timeout_ms'],
                enable_idempotence=config['enable_idempotence'],
                max_in_flight_requests_per_connection=config['max_in_flight_requests_per_connection'],
                compression_type=config['compression_type'],
                batch_size=config['batch_size'],
                linger_ms=config['linger_ms'],
                buffer_memory=config['buffer_memory'],
                key_serializer=lambda k: k.encode('utf-8') if k else None,
                value_serializer=lambda v: v.encode('utf-8'),
            )
            
            print(f"✅ Connected to Kafka successfully!")
            return True
            
        except Exception as e:
            print(f"❌ Failed to connect to Kafka: {e}")
            return False
    
    def send_message(self, key: Optional[str], message: str) -> bool:
        """
        Send a message to the Kafka topic.
        
        Args:
            key: Message key (for partitioning)
            message: Message content
            
        Returns:
            True if message sent successfully, False otherwise
        """
        try:
            future = self.producer.send(self.topic, key=key, value=message)
            
            # Add callback for delivery confirmation
            future.add_callback(self._on_send_success)
            future.add_errback(self._on_send_error)
            
            return True
            
        except Exception as e:
            print(f"❌ Error sending message: {e}")
            self.stats.record_error()
            return False
    
    def _on_send_success(self, record_metadata):
        """Callback for successful message delivery."""
        self.stats.record_sent()
        print(f"✅ Message sent to {record_metadata.topic}[{record_metadata.partition}] "
              f"at offset {record_metadata.offset}")
    
    def _on_send_error(self, exception):
        """Callback for message delivery failure."""
        self.stats.record_error()
        print(f"❌ Message delivery failed: {exception}")
    
    def produce_messages(self, count: int, rate: float = 1.0):
        """
        Produce a specified number of messages.
        
        Args:
            count: Number of messages to produce
            rate: Messages per second (0 = no limit)
        """
        print(f"🚀 Starting to produce {count} messages to topic '{self.topic}'")
        print(f"📊 Rate limit: {rate} messages/second" if rate > 0 else "📊 Rate limit: unlimited")
        
        sleep_time = 1.0 / rate if rate > 0 else 0
        
        for i in range(count):
            if not self.running:
                print("🛑 Stopping due to shutdown signal")
                break
            
            # Create sample message
            message_data = create_sample_message(i, "producer_demo")
            message_json = serialize_message(message_data)
            message_key = f"key_{i}"
            
            # Send message
            success = self.send_message(message_key, message_json)
            
            if success:
                print(f"📤 Sent message {i+1}/{count}: {message_key}")
            
            # Rate limiting
            if sleep_time > 0:
                time.sleep(sleep_time)
            
            # Print stats every 100 messages
            if (i + 1) % 100 == 0:
                self.stats.print_stats()
        
        # Flush remaining messages
        print("🔄 Flushing remaining messages...")
        if self.producer:
            self.producer.flush(timeout=30)
        
        print(f"✅ Finished producing messages!")
        self.stats.print_stats()
    
    def close(self):
        """Close the producer connection."""
        if self.producer:
            print("🔌 Closing producer connection...")
            self.producer.close(timeout=30)
            print("✅ Producer closed successfully")


def main():
    """Main function to run the Kafka producer."""
    parser = argparse.ArgumentParser(description="Kafka Producer Example")
    parser.add_argument(
        "--topic", 
        default=get_default_topic(),
        help=f"Kafka topic to produce to (default: {get_default_topic()})"
    )
    parser.add_argument(
        "--count", 
        type=int, 
        default=10,
        help="Number of messages to produce (default: 10)"
    )
    parser.add_argument(
        "--rate", 
        type=float, 
        default=1.0,
        help="Messages per second (0 for unlimited, default: 1.0)"
    )
    
    args = parser.parse_args()
    
    print("🎯 Kafka Producer Example")
    print("=" * 50)
    print(f"📌 Topic: {args.topic}")
    print(f"📊 Message count: {args.count}")
    print(f"⚡ Rate: {args.rate} messages/second")
    print("=" * 50)
    
    producer = EnhancedKafkaProducer(args.topic)
    
    try:
        # Connect to Kafka
        if not producer.connect():
            sys.exit(1)
        
        # Produce messages
        producer.produce_messages(args.count, args.rate)
        
    except KeyboardInterrupt:
        print("\n🛑 Interrupted by user")
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)
    finally:
        producer.close()


if __name__ == "__main__":
    main()
